File: crates/simcity_sim/src/game/buildings/decay.rs
use bevy::ecs::message::MessageReader;
use bevy::prelude::*;

use crate::game::economy::EconomyConfig;
use crate::game::map::{DirtyTiles, MapGrid, TilePos};
use crate::game::sim::City;
use crate::game::sim_events::DayAdvanced;

use super::components::*;
use super::footprint::{all_footprint_tiles, any_footprint_tile, for_each_footprint_tile};
use super::visual::BuildingTint;

pub fn despawn_invalid_buildings(
    mut commands: Commands,
    mut grid: ResMut<MapGrid>,
    mut dirty: ResMut<DirtyTiles>,
    q: Query<(Entity, &Building)>,
) {
    // A despawn MUST release the building's grid cells: an entity-less cell with
    // `building` still set is a phantom tile that permanently blocks zoning and
    // placement (both validators reject occupied cells) and has no owner the
    // whole-building erase could find.
    for (e, b) in &q {
        // Check if anchor position is valid
        let Some(cell) = grid.get(b.anchor_pos) else {
            clear_footprint_cells(&mut grid, &mut dirty, b);
            commands.entity(e).despawn();
            continue;
        };
        let expected_zone = b.kind.as_zone();
        // Check if anchor cell matches expected building
        if cell.water || cell.zone != expected_zone || cell.building != Some(b.kind) {
            clear_footprint_cells(&mut grid, &mut dirty, b);
            commands.entity(e).despawn();
            continue;
        }
        // Check all footprint tiles are still valid
        let all_valid = all_footprint_tiles(
            b.anchor_pos,
            b.footprint_width,
            b.footprint_length,
            |tile| {
                grid.get(tile)
                    .is_some_and(|cell| !cell.water && cell.building == Some(b.kind))
            },
        );
        if !all_valid {
            clear_footprint_cells(&mut grid, &mut dirty, b);
            commands.entity(e).despawn();
        }
    }
}

/// Clear the building layer of `b`'s footprint cells (only those still owned
/// by its kind — a cell already overwritten by another building is left alone).
fn clear_footprint_cells(grid: &mut MapGrid, dirty: &mut DirtyTiles, b: &Building) {
    for_each_footprint_tile(
        b.anchor_pos,
        b.footprint_width,
        b.footprint_length,
        |tile| {
            if let Some(mut cell) = grid.get(tile)
                && cell.building == Some(b.kind)
            {
                cell.building = None;
                grid.set(tile, cell);
                if let Some(idx) = grid.idx(tile) {
                    dirty.mark(idx);
                }
            }
        },
    );
}

/// Insert/remove the decay warning tint only when it actually changes —
/// re-inserting an identical tint every day would retrigger
/// `Changed<BuildingTint>` and make the render side rebuild for nothing.
fn set_tint(
    commands: &mut Commands,
    e: Entity,
    current: Option<&BuildingTint>,
    want: Option<Color>,
) {
    match (current, want) {
        (Some(c), Some(w)) if c.0 == w => {}
        (None, None) => {}
        (_, Some(w)) => {
            commands.entity(e).insert(BuildingTint(w));
        }
        (Some(_), None) => {
            commands.entity(e).remove::<BuildingTint>();
        }
    }
}

/// GDD: Buildings without road access are demolished after 1 game day
#[allow(clippy::type_complexity, clippy::too_many_arguments)]
pub fn building_decay_no_road_access(
    mut day_events: MessageReader<'_, '_, DayAdvanced>,
    mut commands: Commands,
    mut grid: ResMut<MapGrid>,
    mut dirty: ResMut<DirtyTiles>,
    city: Res<City>,
    mut q: Query<(
        Entity,
        &Building,
        Option<&NoRoadAccessDecay>,
        Option<&BuildingTint>,
    )>,
) {
    // Check if any days advanced (only process decay on day changes)
    let _days_advanced = day_events.read().count();
    let current_day = city.day;

    for (e, b, decay, sprite) in q.iter_mut() {
        // Check if any tile in footprint has road access
        let has_access = any_footprint_tile(
            b.anchor_pos,
            b.footprint_width,
            b.footprint_length,
            |tile| has_adjacent_road(&grid, tile),
        );

        if has_access {
            if decay.is_some() {
                commands.entity(e).remove::<NoRoadAccessDecay>();
            }
            // Restore normal color for service buildings
            if matches!(
                b.kind,
                crate::game::map::BuildingKind::FireStation
                    | crate::game::map::BuildingKind::PoliceStation
                    | crate::game::map::BuildingKind::Hospital
            ) {
                set_tint(&mut commands, e, sprite, None);
            }
            continue;
        }

        // No access: track when access was lost and check if grace period expired
        let access_lost_day = decay
            .as_ref()
            .map(|d| d.access_lost_day)
            .unwrap_or(current_day); // If decay component doesn't exist, start tracking now

        // Check if grace period expired (GDD: 1 game day)
        let days_without_access = current_day.saturating_sub(access_lost_day);

        // Only add/update component if not expired yet (avoid adding component to entity we're about to despawn)
        if days_without_access < NO_ROAD_ACCESS_GRACE_DAYS {
            // Verify building still exists in grid before adding component
            // (another system might have despawned it)
            let building_still_valid = grid
                .get(b.anchor_pos)
                .map(|cell| {
                    cell.building == Some(b.kind) && !cell.water && cell.zone == b.kind.as_zone()
                })
                .unwrap_or(false);

            if !building_still_valid {
                // Building was despawned by another system, skip
                continue;
            }

            // GDD 10.5.1: Visual indicator for service buildings without road access
            if matches!(
                b.kind,
                crate::game::map::BuildingKind::FireStation
                    | crate::game::map::BuildingKind::PoliceStation
                    | crate::game::map::BuildingKind::Hospital
            ) {
                // Change color to red to indicate problem (GDD: visual marking for player)
                set_tint(&mut commands, e, sprite, Some(Color::srgb(1.0, 0.3, 0.3)));
            }

            // Add decay component if not present
            if decay.is_none() {
                commands.entity(e).insert(NoRoadAccessDecay {
                    access_lost_day: current_day,
                });
            }
            continue;
        }

        // Grace period expired - demolish the building
        // GDD 10.5.1: Visual indicator for service buildings without road access (already shown above)
        if matches!(
            b.kind,
            crate::game::map::BuildingKind::FireStation
                | crate::game::map::BuildingKind::PoliceStation
                | crate::game::map::BuildingKind::Hospital
        ) {
            // Keep red color as warning (building will be demolished)
            set_tint(&mut commands, e, sprite, Some(Color::srgb(1.0, 0.3, 0.3)));
        }

        // Demolish: remove from sim state and despawn entity.
        // Clear all footprint tiles
        for_each_footprint_tile(
            b.anchor_pos,
            b.footprint_width,
            b.footprint_length,
            |tile| {
                if let Some(mut cell) = grid.get(tile)
                    && cell.building == Some(b.kind)
                {
                    cell.building = None;
                    grid.set(tile, cell);
                    if let Some(idx) = grid.idx(tile) {
                        dirty.mark(idx);
                    }
                }
            },
        );

        // Population is now calculated from occupancy, not subtracted here
        // The occupancy system will handle population changes

        commands.entity(e).despawn();
    }
}

fn has_adjacent_road(grid: &MapGrid, pos: TilePos) -> bool {
    let neighbors = [
        TilePos {
            x: pos.x - 1,
            y: pos.y,
        },
        TilePos {
            x: pos.x + 1,
            y: pos.y,
        },
        TilePos {
            x: pos.x,
            y: pos.y - 1,
        },
        TilePos {
            x: pos.x,
            y: pos.y + 1,
        },
    ];

    for npos in neighbors {
        if let Some(cell) = grid.get(npos)
            && cell.road.is_some()
        {
            return true;
        }
    }
    false
}

/// Track and process building decay due to low happiness.
///
/// Buildings with average happiness below LOW_HAPPINESS_THRESHOLD for LOW_HAPPINESS_GRACE_DAYS
/// become abandoned and are demolished.
#[allow(clippy::type_complexity, clippy::too_many_arguments)]
pub fn building_decay_low_happiness(
    mut day_events: MessageReader<'_, '_, DayAdvanced>,
    mut commands: Commands,
    mut grid: ResMut<MapGrid>,
    mut dirty: ResMut<DirtyTiles>,
    city: Res<City>,
    economy_cfg: Res<EconomyConfig>,
    mut q: Query<(
        Entity,
        &Building,
        Option<&LowHappinessDecay>,
        Option<&BuildingTint>,
    )>,
) {
    let _days_advanced = day_events.read().count();
    let current_day = city.day;

    // Calculate target happiness based on economy config
    let _target_happiness = economy_cfg.happiness_target;

    for (e, b, decay, sprite) in q.iter_mut() {
        // Only residential and commercial buildings have happiness
        if !matches!(
            b.kind,
            crate::game::map::BuildingKind::Residential
                | crate::game::map::BuildingKind::Commercial
        ) {
            continue;
        }

        // Skip buildings under construction
        if !b.is_operational() {
            continue;
        }

        // Calculate approximate happiness based on occupancy vs target
        // This is a simplified model - real happiness would come from citizen simulation
        let occupancy_ratio = if b.kind == crate::game::map::BuildingKind::Residential {
            b.occupancy_residents as f32 / b.capacity_residents as f32
        } else {
            b.occupancy_jobs as f32 / b.capacity_jobs as f32
        };

        // Happiness is high when occupancy is close to target
        let estimated_happiness = occupancy_ratio.clamp(0.0, 1.0);

        if estimated_happiness >= LOW_HAPPINESS_THRESHOLD {
            // Happiness recovered - remove decay component
            if decay.is_some() {
                commands.entity(e).remove::<LowHappinessDecay>();
                // Restore normal sprite color
                {
                    set_tint(&mut commands, e, sprite, None);
                }
            }
            continue;
        }

        // Low happiness detected
        let decay_start_day = decay
            .as_ref()
            .map(|d| d.decay_start_day)
            .unwrap_or(current_day);

        let days_with_low_happiness = current_day.saturating_sub(decay_start_day);

        // Visual indicator: yellow/orange tint for struggling buildings
        if days_with_low_happiness < LOW_HAPPINESS_GRACE_DAYS {
            {
                // Yellow tint to indicate problems
                set_tint(&mut commands, e, sprite, Some(Color::srgb(1.0, 0.8, 0.3)));
            }

            // Add decay component if not present
            if decay.is_none() {
                commands.entity(e).insert(LowHappinessDecay {
                    decay_start_day: current_day,
                    avg_happiness: estimated_happiness,
                });
            }
            continue;
        }

        // Grace period expired - abandon building
        demolish_building(e, b, &mut commands, &mut grid, &mut dirty);
    }
}

/// Track and process building decay due to economic losses.
///
/// Buildings with cumulative losses below ECONOMIC_LOSSES_THRESHOLD are abandoned.
#[allow(clippy::type_complexity, clippy::too_many_arguments)]
pub fn building_decay_economic(
    mut day_events: MessageReader<'_, '_, DayAdvanced>,
    mut commands: Commands,
    mut grid: ResMut<MapGrid>,
    mut dirty: ResMut<DirtyTiles>,
    city: Res<City>,
    mut q: Query<(
        Entity,
        &Building,
        Option<&EconomicDecay>,
        Option<&BuildingTint>,
    )>,
) {
    let _days_advanced = day_events.read().count();
    let current_day = city.day;

    for (e, b, decay, sprite) in q.iter_mut() {
        // Only commercial and industrial buildings have economic concerns
        if !matches!(
            b.kind,
            crate::game::map::BuildingKind::Commercial | crate::game::map::BuildingKind::Industrial
        ) {
            continue;
        }

        // Skip buildings under construction
        if !b.is_operational() {
            continue;
        }

        // Calculate economic performance based on occupancy
        // Low occupancy = low income = losses
        let occupancy_ratio = b.occupancy_jobs as f32 / b.capacity_jobs as f32;

        // Estimate daily loss based on low occupancy
        // This is simplified - real economy would track actual money flow
        let estimated_daily_loss = if occupancy_ratio < 0.5 {
            -(((0.5 - occupancy_ratio) * 20.0) as i64) // Up to -10 per day
        } else {
            0
        };

        if estimated_daily_loss == 0 {
            // Economic recovery
            if decay.is_some() {
                commands.entity(e).remove::<EconomicDecay>();
                {
                    set_tint(&mut commands, e, sprite, None);
                }
            }
            continue;
        }

        let decay_start_day = decay
            .as_ref()
            .map(|d| d.decay_start_day)
            .unwrap_or(current_day);

        let days_with_losses = current_day.saturating_sub(decay_start_day);
        let cumulative_losses = (days_with_losses as i64) * estimated_daily_loss;

        if cumulative_losses > ECONOMIC_LOSSES_THRESHOLD {
            // Add/update decay component
            if decay.is_none() {
                commands.entity(e).insert(EconomicDecay {
                    decay_start_day: current_day,
                    cumulative_losses,
                });
            }

            // Visual indicator: purple tint for economic trouble
            {
                set_tint(&mut commands, e, sprite, Some(Color::srgb(0.8, 0.5, 1.0)));
            }
            continue;
        }

        // Economic losses exceeded threshold - abandon building
        demolish_building(e, b, &mut commands, &mut grid, &mut dirty);
    }
}

/// Helper function to demolish a building
fn demolish_building(
    e: Entity,
    b: &Building,
    commands: &mut Commands,
    grid: &mut ResMut<MapGrid>,
    dirty: &mut ResMut<DirtyTiles>,
) {
    // Clear all footprint tiles
    for_each_footprint_tile(
        b.anchor_pos,
        b.footprint_width,
        b.footprint_length,
        |tile| {
            if let Some(mut cell) = grid.get(tile)
                && cell.building == Some(b.kind)
            {
                cell.building = None;
                grid.set(tile, cell);
                if let Some(idx) = grid.idx(tile) {
                    dirty.mark(idx);
                }
            }
        },
    );

    commands.entity(e).despawn();
}

File: packages/sim/src/buildings/decay.ts
// Port of crates/simcity_sim/src/game/buildings/decay.rs: buildings the grid no longer backs, and buildings
// abandoned for lost road access, unhappy occupants or economic losses. The warning tints are render-side.
import { crimeContentment } from '../cityFields';
import { zoneDemand } from '../demand';
import { bumpVersion } from '../map/dirty';
import type { World } from '../world';
import {
  allFootprintTiles,
  anyFootprintTile,
  buildingArea,
  buildingKindZone,
  constructionDays,
  footprintTiles,
  hasAdjacentRoad,
  isOperational,
  type Building,
} from './building';
import { expectedOccupancy } from './occupancy';

const f32 = Math.fround;

/** GDD: a building without road access is demolished after a game day. */
export const NO_ROAD_ACCESS_GRACE_DAYS = 1;
/** GDD: days of low happiness before a building is abandoned. */
export const LOW_HAPPINESS_GRACE_DAYS = 2;
/** Happiness below which decay starts. */
export const LOW_HAPPINESS_THRESHOLD = f32(0.3);
/** GDD: losses at which a building is abandoned. */
export const ECONOMIC_LOSSES_THRESHOLD = -100;

/**
 * Removes `b` and clears the building layer of its footprint where it still shows `b`'s kind: a cell left
 * with a building and no record would block zoning and placement for good. The map edit version bumps so
 * the renderer and the utility network see the change (Rust left the version alone).
 */
export function demolishBuilding(w: World, b: Building): void {
  for (const tile of footprintTiles(b)) {
    const cell = w.grid.get(tile);
    if (cell === undefined || cell.building !== b.kind) continue;
    w.grid.set(tile, { ...cell, building: null });
    w.dirty.mark(w.grid.idx(tile)!);
  }
  w.buildings.remove(b.id);
  w.mapEditVersion = bumpVersion(w.mapEditVersion);
}

/** `despawn_invalid_buildings`: a building whose anchor or footprint the grid no longer shows goes. */
export function despawnInvalidBuildings(w: World): void {
  const grid = w.grid;
  for (const b of [...w.buildings.all()]) {
    const cell = grid.get(b.anchor);
    const valid =
      cell !== undefined &&
      !cell.water &&
      cell.zone === buildingKindZone(b.kind) &&
      cell.building === b.kind &&
      allFootprintTiles(b.anchor, b.width, b.length, (tile) => {
        const c = grid.get(tile);
        return c !== undefined && !c.water && c.building === b.kind;
      });
    if (!valid) demolishBuilding(w, b);
  }
}

/** `building_decay_no_road_access`: without a road beside its footprint a building is demolished after the grace day. */
export function buildingDecayNoRoadAccess(w: World): void {
  const day = w.city.day;
  for (const b of [...w.buildings.all()]) {
    if (anyFootprintTile(b.anchor, b.width, b.length, (tile) => hasAdjacentRoad(w.grid, tile))) {
      b.noRoadAccessDecay = null;
      continue;
    }
    const lostDay = b.noRoadAccessDecay?.accessLostDay ?? day;
    if (Math.max(day - lostDay, 0) < NO_ROAD_ACCESS_GRACE_DAYS) {
      b.noRoadAccessDecay ??= { accessLostDay: day };
      continue;
    }
    demolishBuilding(w, b);
  }
}

/**
 * `building_decay_low_happiness`: homes and shops whose happiness, approximated from occupancy against what
 * the market lets them hold by now and from crime, stays low past the grace days are abandoned.
 */
export function buildingDecayLowHappiness(w: World): void {
  const day = w.city.day;
  for (const b of [...w.buildings.all()]) {
    if ((b.kind !== 'Residential' && b.kind !== 'Commercial') || !isOperational(b)) continue;
    const homes = b.kind === 'Residential';
    const occupancy = homes ? b.occupancyResidents : b.occupancyJobs;
    const target = homes ? b.targetOccupancyResidents : b.targetOccupancyJobs;
    let ratio: number;
    if (target === 0) {
      // The market gives it nobody: no road, no power, or demand has collapsed.
      ratio = 0;
    } else {
      const area = buildingArea(b);
      const openedDay = b.constructionStartDay + constructionDays(b.kind, b.level, area);
      const expected = expectedOccupancy(target, b.level, area, zoneDemand(w.rciDemand, b.kind), Math.max(day - openedDay, 0));
      ratio = expected < 1 ? 1 : f32(occupancy / expected);
    }
    const crime = homes ? w.cityFields.footprintMean('Crime', w.grid, b.anchor, b.width, b.length) : undefined;
    const happiness = f32(Math.min(Math.max(ratio, 0), 1) * (crime === undefined ? 1 : crimeContentment(crime)));
    if (happiness >= LOW_HAPPINESS_THRESHOLD) {
      b.lowHappinessDecay = null;
      continue;
    }
    const startDay = b.lowHappinessDecay?.decayStartDay ?? day;
    if (Math.max(day - startDay, 0) < LOW_HAPPINESS_GRACE_DAYS) {
      b.lowHappinessDecay ??= { decayStartDay: day, avgHappiness: happiness };
      continue;
    }
    demolishBuilding(w, b);
  }
}

/** `building_decay_economic`: shops and industry below half their jobs lose money daily and are abandoned past the threshold. */
export function buildingDecayEconomic(w: World): void {
  const day = w.city.day;
  for (const b of [...w.buildings.all()]) {
    if ((b.kind !== 'Commercial' && b.kind !== 'Industrial') || !isOperational(b)) continue;
    const ratio = f32(b.occupancyJobs / b.capacityJobs);
    const dailyLoss = ratio < 0.5 ? -Math.trunc(f32(f32(0.5 - ratio) * 20)) : 0;
    if (dailyLoss === 0) {
      b.economicDecay = null;
      continue;
    }
    const startDay = b.economicDecay?.decayStartDay ?? day;
    const losses = Math.max(day - startDay, 0) * dailyLoss;
    if (losses > ECONOMIC_LOSSES_THRESHOLD) {
      b.economicDecay ??= { decayStartDay: day, cumulativeLosses: losses };
      continue;
    }
    demolishBuilding(w, b);
  }
}
